Toric/Mathlib/Geometry/Convex/Cone/Dual.lean

```diff
@@ -0,0 +1,91 @@
+
+# The algebraic dual of a cone
+
+Given a bilinear pairing `p` between two `R`-modules `M` and `N` and a set `s` in `M`, we define
+`PointedCone.dual p C` to be the pointed cone in `N` consisting of all points `y` such that for all
+`x ∈ s` we have `0 ≤ p x y`.
+
+When the pairing is perfect, this gives us the algebraic dual of a cone. This is developed here.
+When the pairing is continuous and perfect (as a continuous pairing), this gives us the topological
+dual instead. See `Mathlib.Analysis.Convex.Cone.Dual` for that case.
+-/
+
+open Function LinearMap Pointwise Set
+
+namespace PointedCone
+variable {R M N : Type*} [CommRing R] [PartialOrder R] [IsOrderedRing R] [AddCommGroup M]
+  [AddCommGroup N] [Module R M] [Module R N] {p : M →ₗ[R] N →ₗ[R] R} {s t : Set M} {y : N}
+
+variable (p s) in
+/-- The dual cone of a set `s` with respect to a bilinear pairing `p` is the cone consisting of all
+points `y` such that for all points `x ∈ s` we have `0 ≤ p x y`. -/
+def dual' (s : Set M) : PointedCone R N where
+  carrier := {y | ∀ ⦃x⦄, x ∈ s → 0 ≤ p x y}
+  zero_mem' := by simp
+  add_mem' {u v} hu hv x hx := by rw [map_add]; exact add_nonneg (hu hx) (hv hx)
+  smul_mem' c y hy x hx := by rw [← Nonneg.coe_smul, map_smul]; exact mul_nonneg c.2 (hy hx)
+
+@[simp] lemma mem_dual' : y ∈ dual' p s ↔ ∀ ⦃x⦄, x ∈ s → 0 ≤ p x y := .rfl
+
+@[simp] lemma dual_empty : dual' p ∅ = ⊤ := by ext; simp
+@[simp] lemma dual_zero : dual' p 0 = ⊤ := by ext; simp
+
+lemma dual_univ (hp : Injective p.flip) : dual' p univ = 0 := by
+  refine le_antisymm (fun y hy ↦ (_root_.map_eq_zero_iff p.flip hp).1 ?_) (by simp)
+  ext x
+  exact (hy <| mem_univ x).antisymm' <| by simpa using hy <| mem_univ (-x)
+
+@[gcongr] lemma dual_le_dual (h : t ⊆ s) : dual' p s ≤ dual' p t := fun _y hy _x hx ↦ hy (h hx)
+
+lemma dual_union (s t : Set M) : dual' p (s ∪ t) = dual' p s ⊓ dual' p t := by aesop
+
+lemma dual_insert (x : M) (s : Set M) : dual' p (insert x s) = dual' p {x} ⊓ dual' p s := by
+  rw [insert_eq, dual_union]
+
+lemma dual_iUnion {ι : Sort*} (f : ι → Set M) : dual' p (⋃ i, f i) = ⨅ i, dual' p (f i) := by
+  ext; simp [forall_swap (α := M)]
+
+lemma dual_sUnion (S : Set (Set M)) : dual' p (⋃₀ S) = sInf (dual' p '' S) := by
+  ext; simp [forall_swap (α := M)]
+
+/-- The dual cone of `s` equals the intersection of dual cones of the points in `s`. -/
+lemma dual_eq_iInter_dual_singleton (s : Set M) :
+    dual' p s = ⋂ i : s, (dual' p {i.val} : Set N) := by
+  ext; simp [forall_swap (α := M)]
+
+/-- Any set is a subset of its double dual cone. -/
+lemma subset_dual_dual : s ⊆ dual' p.flip (dual' p s) := fun _x hx _y hy ↦ hy hx
+
+-------------------------------------------------------------
+-- Things below this line are not part of mathlib's #24149 --
+-------------------------------------------------------------
+
+@[simp]
+lemma dual_dual_dual_eq_dual : dual' p (dual' p.flip (dual' p s)) = dual' p s :=
```

Comment:
```suggestion
lemma dual_dual_flip_dual_eq_dual : dual' p (dual' p.flip (dual' p s)) = dual' p s :=
```
and can we also acquire `dual_flip_dual_dual_flip_eq_dual_flip`?

Reply:
I added `dual_flip_dual_dual_flip_eq_dual_flip`. Note that it is defeq with `dual_dual_flip_dual_eq_dual` however, since `p.flip.flip` is defeq with `p`.

Reply:
Yep, that's correct. It is *syntactically* different, which is why we need it